Add Christmas blog post, 3.13 to 3.21 (#199)
transloadit/uppy.io: blog/2023-12-3.13-3.21.md

```

:::tip The old way via `.use(ImageEditor, { target: Dashboard })` and
`.use(Dashboard, { plugins: ['Webcam', 'GoogleDrive'] })` still works, but we
encourage you to use the new syntax :::

Bug fixes:

- Remove `.uppy-Dashboard-isFixed` CSS class when `uppy.close()` is invoked.
- When `showAddFilesPanel: true`, `aria-hidden` should be the opposite.
- Don’t cancel all files and Assemblies when clicking “done”.
- Update Dashboard’s blue color to meet WCAG contrast requirements.

## AWS S3

- Fixed `TypeError`, fixed `uploadURL` when using `PUT`.
- Pass `signal` as separate arg for backward compat; pass the `uploadURL` back
  to the caller.
- Refresh file before calling user-defined functions.
- Made sure we retry signature request.
- Use uppercase HTTP method names.
- Call `#setCompanionHeaders` in `setOptions`.

## Companion

One of the most notable changes in Companion is on the client side, we’ve made
authentication optional, allowing for more provider types, with publicly
accessable links, for example.

In Companion Client we were able to drop the now unnecessary preflight requests,
improving upload speeds and latency yet again.

Other notable changes and fixes:

> - Allow dynamic S3 bucket

Reviewer: All list items here are sentences, and therefore should end with a dot

It’s this time of the year again! Make yourself cozy
[by the fire](https://www.youtube.com/watch?v=ouqtX6WWL1U), and let us help you
unwrap the gifts Uppy has prepared for you.

We proud of all the improvements and amazing new features and bug fixes we were
able to introduce this year. The highlights being: conditional S3 multipart and
signing S3 uploads on the client, speedy handling of 10-50k file uploads,
refresh tokens (useful when the token expires mid-upload for a large amount of
files / slow connections), server-sent events and new `assemblyOptions` for the
Transloadit plugin, Dashboard Single File mode (`singleFileFullScreen`),
improved ETA in the Status Bar, a bunch of new languages, including Hindi,
Mexican Spanish and Catalan.

This one, however, is a recent Uppy updates post, covering releases from
[`3.13.0` to `3.21.0`](https://github.com/transloadit/uppy/compare/uppy%403.13.0...uppy%403.21.0):
announcing the switch to Typescript, Dashboard can now auto-install plugins,
support for Catalan language, a heap of bug fixes for Tus, Companion, AWS S3,
and more.

<!--truncate-->

## Typescript

> We are excited to announce that we are working hard to add Typescript types to
> Uppy Core and all of the plugins. We have a build setup in place that

Reviewer: Technically there were already types
```suggestion
We are excited to announce that we are working hard to refactor
Uppy Core and all of the plugins source files to TypeScript. We have a build setup in place that
```

temporarily supports both `.ts` and `.js` files in our monorepo. And a helper
utility that takes a package name and does the initial renaming and Typescript
config. Then we go through the list of plugins one by one. We begun with
`@uppy/utils` and `@uppy/core`, since most other packages depend on those. Now
we are moving to the rest, one by one.

This work so far is only affecting Uppy internals. In published packages we are
using the manually crafted `d.ts` files, as before. We are planning to flip the
switch and introduce some breaking changes in the next Uppy major, due to be
released in the first quater of 2024.

## Core

> - Allow duplicate files with `onBeforeFileAdded`
> - Fixed`onBeforeFileAdded` with Golden Retriever
> - Simplify types with class generic and type more events
> - Export `UIPlugin` and `BasePlugin` in Uppy CDN bundle

Reviewer: ```suggestion
- Allow duplicate files with `onBeforeFileAdded`.
- Fixed`onBeforeFileAdded` with Golden Retriever.
- Simplify types with class generic and type more events.
- Export `UIPlugin` and `BasePlugin` in Uppy CDN bundle.
```

> - Export `UIPlugin` and `BasePlugin` in Uppy CDN bundle

Reviewer: I wonder if we want to put emphasis on that, we wouldn't want a popular Uppy community plugin to be using those

## Image Editor

We have a whole separate blog post about the recent rotation improvements in the
Image Editor, so please read that, if you are interested. Other changes include:

> - Fixed labels on all buttons, centered the rotation slider label
> - Limited cropbox movements
> - Fixed to respect `cropperOptions.initialAspectRatio`
> - Change the checkered background color, make it semi-transparent
> - Fixed granular rotation: `[-45, 45]` instead of `[-45, 44]`
> - Removed 1px black lines
> - Made `file-editor:cancel` event fire when the Image Editor’s “cancel” button
>   is pressed

Reviewer: ```suggestion
- Fixed labels on all buttons, centered the rotation slider label.
- Limited cropbox movements.
- Fixed to respect `cropperOptions.initialAspectRatio`.
- Change the checkered background color, make it semi-transparent.
- Fixed granular rotation: `[-45, 45]` instead of `[-45, 44]`.
- Removed 1px black lines.
- Made `file-editor:cancel` event fire when the Image Editor’s “cancel” button
  is pressed.
```

> <video muted autoplay loop>
>   <source src="/img/blog/3.13-3.21/crop.mov" type="video/mp4" />
> </video>

Reviewer: nit: you technically don't need a `<source>` if you have only one
```suggestion
<video muted autoplay loop src="/img/blog/3.13-3.21/crop.mov" type="video/mp4">
</video>
```

Author: For sure, I just copy-pasted from previous post.

<img alt="Editor semi-transparant checkered background" src="/img/blog/3.13-3.21/editor.jpg" />

## Dashboard

Dashboard can now automatically discover and install compatible plugins without

> `target` option. Here’s how it works:

Reviewer: I don't think that conde syntax shows how it works, a more correct intro would be something like
```suggestion
`target` option. This lets you use a much more condense syntax:
```

```js
const uppy = new Uppy()
	.use(ImageEditor)
	.use(Dashboard)
	.use(Webcam)
	.use(GoogleDrive);
```

:::tip The old way via `.use(ImageEditor, { target: Dashboard })` and

> `.use(Dashboard, { plugins: ['Webcam', 'GoogleDrive'] })` still work, but might

Reviewer: ```suggestion
`.use(Dashboard, { plugins: ['Webcam', 'GoogleDrive'] })` still works, but might
```

> be deprecated in the next major. :::

Reviewer: I think we should either say _why_ we would be deprecating it, or not say it at all.

Bug fixes:

> - Remove `.uppy-Dashboard-isFixed` CSS class when `uppy.close()` is invoked
> - When `showAddFilesPanel: true`, `aria-hidden` should be the opposite
> - Don’t cancel all files and Assemblies when clicking “done”
> - Update Dashboard’s blue color to meet WCAG contrast requirements

Reviewer: ```suggestion
- Remove `.uppy-Dashboard-isFixed` CSS class when `uppy.close()` is invoked.
- When `showAddFilesPanel: true`, `aria-hidden` should be the opposite.
- Don’t cancel all files and Assemblies when clicking “done”.
- Update Dashboard’s blue color to meet WCAG contrast requirements.
```

## AWS S3

> - Fixed `TypeError`, fixed `uploadURL` when using `PUT`
> - Pass `signal` as separate arg for backward compat; pass the `uploadURL` back
>   to the caller
> - Refresh file before calling user-defined functions
> - Made sure we retry signature request
> - Use uppercase HTTP method names
> - Call `#setCompanionHeaders` in `setOptions`

Reviewer: ```suggestion
- Fixed `TypeError`, fixed `uploadURL` when using `PUT`.
- Pass `signal` as separate arg for backward compat; pass the `uploadURL` back
  to the caller.
- Refresh file before calling user-defined functions.
- Made sure we retry signature request.
- Use uppercase HTTP method names.
- Call `#setCompanionHeaders` in `setOptions`.
```
